checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -362,57 +398,128 @@ private static int getGroupCount(Pattern p) {
+                                openingIndices.push(group);
+                                openingWasGroup.push(true);
+                            } else { // non capturing group.
+                                openingWasGroup.push(false);
+                            }
+                        } else { // unnamed capturing group.
+                            group += 1;
+                            if (i != 0 && regexp.charAt(i - 1) == '|')
+                                nonNullGroups.remove(Integer.valueOf(group));
+                            openingIndices.push(group);
+                            openingWasGroup.push(true);
+                        }
+                    }
+                } else if (escaped) { // escaped won't be true if quoted is true.
+                    escaped = false;
                 }
-                group += 1;
-                if (i != 0 && regexp.charAt(i - 1) == '|') {
-                    nonNullGroups.remove((Integer) group);
+            } else if (regexp.charAt(i) == ')') {
+                if (!escaped && insideList == 0 && !quoted) { // ending of a construct.
+                    boolean closesGroup = openingWasGroup.pop();
+                    if (closesGroup) {
+                        int value = openingIndices.pop();
+                        if (i != regexp.length() - 1
+                                && "?|*".contains(Character.toString(regexp.charAt(i + 1)))) {
+                            nonNullGroups.remove(Integer.valueOf(value));
+                        } else if (i < length - 2
+                                && regexp.charAt(i + 1) == '{'
+                                && regexp.charAt(i + 2) == '0') {
+                            nonNullGroups.remove(Integer.valueOf(value));
+                        }
+                    }
+                } else if (escaped) {
+                    escaped = false;
                 }
-                openingIndices.push(group);
-            } else if (!escaped && !squareBracketOpen && regexp.charAt(i) == ')') {
-                if (notCapturing) {
-                    notCapturing = false;
-                    continue;
+            } else if (regexp.charAt(i) == '\\') {
+                if (!quoted) {
+                    escaped =
+                            !escaped; // if it is already escaped, '\' will be matched, if not, now
+                    // escaped becomes true.
+                } else if (i != length - 1 && regexp.charAt(i + 1) == 'E') {
+                    escaped = true;
                 }
-                int popped = openingIndices.pop();
-                if (i != regexp.length() - 1
-                        && "*?|".contains(Character.toString(regexp.charAt(i + 1)))) {
-                    nonNullGroups.remove((Integer) popped);
+            } else if (regexp.charAt(i) == '[') {
+                if (!escaped && !quoted) {
+                    insideList += 1;
```

Comment:
Rather than setting a variable and returning to the main loop that tries to handle all sorts of features, I think it would be simpler and easier to understand if you write a little loop here that increments `i` until the matching `]` is found.  Then there is no need for the `regexp.charAt(i) == ']'` test, or the `insideList` variable.